File: python/spiral-matrix/spiral_matrix.py
"""
Generate a grid of size x size and fill it with a spiral matrix starting at the top left and 
filling numbers clockwise toward the center.
"""

DIR_RIGHT = 0
DIR_DOWN = 1
DIR_LEFT = 2
DIR_UP = 3

delta = [(1, 0), (0, 1), (-1, 0), (0, -1)]

def spiral_matrix(size):
    """
    Generate a spiral matrix of size x size starting at the top left and spiraling clockwise
    :param size: The size of the matrix to return
    :returns: a size x size list of lists filled with a spiral matrix
    """
    if size == 0:
        return []
    
    ret = [list(range(size)) for index in range(size)]
    top = 0
    bottom = size - 1
    left = 0
    right = size - 1
    coord_x = 0
    coord_y = top
    delta_dir = DIR_RIGHT
    for counter in range(size * size):
        ret[coord_y][coord_x] = counter + 1

        coord_x += delta[delta_dir][0]
        coord_y += delta[delta_dir][1]

        if coord_x > right:
            # Turn right moving down
            top += 1
            coord_x = right
            coord_y = top
            delta_dir = DIR_DOWN
        elif coord_y > bottom:
            # Turn right moving left
            right -= 1
            coord_x = right
            coord_y = bottom
            delta_dir = DIR_LEFT
        elif coord_x < left:
            # Turn right moving up
            bottom -= 1
            coord_x = left
            coord_y = bottom 
            delta_dir = DIR_UP
        elif coord_y < top:
            left += 1
            coord_x = left
            coord_y = top
            delta_dir = DIR_RIGHT

    return ret

File: go/spiral-matrix/spiral_matrix.go
// Spiral Matrix exercise
package spiralmatrix

// Generate a spiral clockwise matrix of size x size.
// @param size: The width and height of the matrix.
// @returns: size x size matrix with a spiral going clockwise starting at 0, 0
func SpiralMatrix(size int) [][]int {
	ret := make([][]int, size)
	for i := 0; i < size; i++ {
		ret[i] = make([]int, size)
		for j := 0; j < size; j++ {
			ret[i][j] = 0
		}
	}
	dx := []int{1, 0, -1, 0}
	dy := []int{0, 1, 0, -1}
	dir := 0
	x := 0
	y := 0
	counter := 0
	for counter < size*size {
		counter++
		ret[y][x] = counter
		y += dy[dir]
		x += dx[dir]
		if x < 0 || x >= size || y < 0 || y >= size || ret[y][x] != 0 {
			y -= dy[dir] // Step back
			x -= dx[dir]
			dir = (dir + 1) % len(dx) // Turn
			y += dy[dir]              // Move in new direction
			x += dx[dir]
		}
	}

	return ret
}
